src/adc_readout/AdcTimeStamp.h

```diff
@@ -78,3 +98,73 @@ class TimeStamp {
   RawTimeStamp CTime{0, 0};
   ClockMode CMode{ClockMode::External};
 };
+
+template <typename ChronoTime>
+TimeStamp MakeExternalTimeStampFromClock(ChronoTime TimeNow) {
+  auto NowSeconds = std::chrono::duration_cast<std::chrono::seconds>(
+                        TimeNow.time_since_epoch())
+                        .count();
+  double NowSecFrac = (std::chrono::duration_cast<std::chrono::nanoseconds>(
+                           TimeNow.time_since_epoch())
+                           .count() /
+                       1e9) -
+                      NowSeconds;
+  std::uint32_t Ticks =
+      std::lround(NowSecFrac * (TimerClockFrequencyExternal / 2.0));
+
+  RawTimeStamp rts{static_cast<uint32_t>(NowSeconds), Ticks};
+  return TimeStamp(rts, TimeStamp::ClockMode::External);
+}
+
+//-----------------------------------------------------------------------------
+
+inline TimeStamp::TimeStamp(std::uint64_t NSec, ClockMode Mode) : CMode(Mode) {
+  CTime.Seconds = NSec / NSecMultiplier;
+  auto NanoSecPart = NSec % NSecMultiplier;
+  auto &CConst = TimeConst[int(CMode)];
+  CTime.SecondsFrac =
+      std::lround(NanoSecPart / TimeConst[int(CMode)].SampleLengthNS);
+  if (CTime.SecondsFrac == CConst.AdcTimerCounterMax) {
```

Comment:
I can't judge if this is robust enough or if it would be better to use >= here.

Reply:
the code is unchanged.